python/ray/experimental/channel/common.py

```diff
@@ -356,7 +381,15 @@ def start(self):
         self._background_task = asyncio.ensure_future(self.run())
 
     def _run(self):
-        return [c.read() for c in self._input_channels]
+        results = []
+        for c in self._input_channels:
+            exiting = retry_and_check_interpreter_exit(
+                lambda: results.append(c.read(timeout=1))
+            )
+            if exiting:
+                break
```

Comment:
Just to understand, the original code hangs because it waits inside C++ and therefore does not respect interpreter exiting?

Reply:
no. it is due to

```
Fix asyncio read/write being blocked and joined forever issue. We check read/write every 1 second and check sys.is_finalizing() which sets to True upon interpreting exiting time. We can't rely on atexit handler.teardown because asyncio read/write runs in thread pool, and thread pool is joined "before at exit handler is executed". See https://github.com/python/cpython/blob/8f82d9aa2191db7826bb7a453fe06ce65f966cf8/Lib/concurrent/futures/thread.py#L37 (this atexit handler always is called before python's regular atexit handler).

```